src/wasm/kotlin/src/arcs/Particle.kt

```diff
@@ -116,94 +116,82 @@ open class Singleton<T : Entity<T>> constructor(val entityCtor: () -> T) : Handl
         entity = entityCtor.invoke().decodeEntity(added)
     }
 
-    fun get(): T? {
-        return entity
-    }
+    fun get() = entity
 
     fun set(entity: T) {
         this.entity = entity
         val encoded = entity.encodeEntity()
         singletonSet(
-            this.particle!!.toWasmAddress(),
+            this.particle.toWasmAddress(),
             this.toWasmAddress(),
             encoded.toWasmString()
         )
     }
 
     fun clear() {
         entity = entityCtor()
-        singletonClear(this.particle!!.toWasmAddress(), this.toWasmAddress())
+        singletonClear(this.particle.toWasmAddress(), this.toWasmAddress())
     }
 }
 
 class Collection<T : Entity<T>> constructor(private val entityCtor: () -> T) : Handle(),
     Iterable<T> {
-    override fun iterator(): Iterator<T> {
-        return entities.values.iterator()
-    }
 
     private val entities: MutableMap<String, T> = mutableMapOf()
 
+    override fun iterator() = entities.values.iterator()
+
     override fun sync(encoded: String) {
         entities.clear()
         add(encoded)
     }
 
     override fun update(added: String, removed: String) {
         add(added)
-        val decoder = StringDecoder(removed)
-            var num = decoder.getInt(":")
+        with (StringDecoder(removed)) {
+            var num = getInt(":")
             while (num-- > 0) {
-                val len = decoder.getInt(":")
-                val chunk = decoder.chomp(len)
+                val len = getInt(":")
+                val chunk = chomp(len)
                 // TODO: just get the id, no need to decode the full entity
                 val entity = entityCtor.invoke().decodeEntity(chunk)!!
                 entities.remove(entity.internalId)
             }
+        }
     }
 
-    fun size(): Int {
-        return entities.size
-    }
-
-    fun empty(): Boolean {
-        return entities.isEmpty()
-    }
+    fun size() = entities.size
+    fun empty() = entities.isEmpty()
 
     fun store(entity: T) {
         entities[entity.internalId] = entity
         val encoded = entities[entity.internalId]!!.encodeEntity()
-        collectionStore(this.particle!!.toWasmAddress(), this.toWasmAddress(), encoded.toWasmString())
+        collectionStore(this.particle.toWasmAddress(), this.toWasmAddress(), encoded.toWasmString())
     }
 
-    fun remove(entity: T) {
-        val it = entities[entity.internalId]
-        if (it != null) {
-            val encoded: String = it.encodeEntity()
-            entities.remove(entity.internalId)
-            collectionRemove(this.particle!!.toWasmAddress(), this.toWasmAddress(), encoded.toWasmString())
-        }
+    fun remove(entity: T) = entities[entity.internalId]?.let {
```

Comment:
This can be a little dangerous.

The return type of `?.let` is a nullable type of the last expression within it.  In this case, if `collectionRemove` is `Unit` (void), then the return type of `remove(entity: T)` is `Unit?`.  That could be a bit ugly.

Reply:
Moved into a regular function block, thanks for pointing out!